core/src/com/unciv/logic/automation/unit/WorkerAutomation.kt

```diff
@@ -118,23 +118,15 @@ class WorkerAutomation(
         localUniqueCache: LocalUniqueCache,
         currentTile: Tile
     ): Boolean {
-        val citiesToNumberOfUnimprovedTiles = HashMap<String, Int>()
-        
-        for (city in unit.civ.cities) {
-            citiesToNumberOfUnimprovedTiles[city.id] = city.getTiles()
-                .count { tile ->
-                    tile.isLand
-                            && tile.getUnits().none { unit -> unit.cache.hasUniqueToBuildImprovements }
-                            && (tile.isPillaged() || tileHasWorkToDo(tile, unit, localUniqueCache))
-                }
-        }
-
-        val closestUndevelopedCity = unit.civ.cities.asSequence()
-            .filter { citiesToNumberOfUnimprovedTiles[it.id]!! > 0 }
-            .sortedByDescending { it.getCenterTile().aerialDistanceTo(currentTile) }
-            .firstOrNull { unit.movement.canReach(it.getCenterTile()) } //goto most undeveloped city
-
-        if (closestUndevelopedCity != null && closestUndevelopedCity != currentTile.owningCity) {
+        // Note, however, that the closest city to a tile isn't necessarily the owning city
+        val closestUndevelopedCity = unit.civ.cities
+            .filter { it != unit.currentTile.owningCity && it.getTiles().any { tile -> tile.isLand
+                    && tile.getUnits().none { unit -> unit.cache.hasUniqueToBuildImprovements }
+                    && (tile.isPillaged() || tileHasWorkToDo(tile, unit, localUniqueCache)) } }
```

Comment:
This is a lot, please move it all into a Tile -> Boolean function

Reply:
It's less lines of code than it used to be, you mean split it up into more lines?

Reply:
It looks ugly imo.